src/core/shape/2d_primitives.js

```diff
@@ -375,7 +375,18 @@ p5.prototype.quad = function() {
   p5._validateParameters('quad', arguments);
 
   if (this._renderer._doStroke || this._renderer._doFill) {
-    this._renderer.quad.apply(this._renderer, arguments);
+    if (this._renderer.isP3D && arguments.length !== 12) {
+      // if 3D and we weren't passed 12 args, assume Z is 0
+      // prettier-ignore
+      this._renderer.quad.call(
+        this._renderer,
+        arguments[0], arguments[1], 0,
+        arguments[2], arguments[3], 0,
+        arguments[4], arguments[5], 0,
+        arguments[6], arguments[7], 0);
+    } else {
+      this._renderer.quad.apply(this._renderer, arguments);
```

Comment:
These changes are great! This works well. My only remaining concern is that really unexpected things happen when you are in 2D and you try to pass 12 arguments. I know this isn't new to your pull request as this problem occurs even with quad as-is with the current release. That said, since you are already doing so much to improve it might as well put in a polite check here and if the rendererer isn't 3D and the argument length is 12 output an informative error to the console. What do you think?

Reply:
ok, I guess its a good idea. I do wonder if we need to do that much checking on behalf of the user.

so have a different if/else before this one? 
or within this last else, just check the arguments, output the error, and call the apply?

Reply:
I tend to err on the side of caution when functions work differently between 2D and 3D. That said I think you are right on this one. If you just add the explanation that z-arguments only work in WEBGL (see comment below) that should be enough. I think using quad instead of rect already implies an understanding of the coordinate system. Thanks for hearing me out!